### PR TITLE
qt => qt5_, add package rename and old package deletion functionality.

### PR DESCRIPTION
**Pull request by satmandu**
- Renames of qt packages to qt5_ since we will need to start working on qt6_ packages at some point.
- Updates of core qt packages to 5.15.11
- Includes logic in `lib/fixup.rb` to rename packages marked for renaming and prompt for deletion of deprecated packages.

Builds properly:
- [x] `x86_64`
- [x] `i686`
- [x] `armv7l` 

### Run the following to get this pull request's changes locally for testing.
```
CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=qt5 CREW_TESTING=1 crew update
```

### CODE REVIEW

**Comment by Zopolis4** [comment]
When you say a TODO pr, do you mean one you have and simply havent submitted yet or one that you plan to make in the future?

I ask because I've also run into the need to rename a few packages in a different project of mine, and I've got some ideas and WIP code on how to do it, so I think it makes sense to avoid doubling up work.

**Comment by satmandu** [comment]
> When you say a TODO pr, do you mean one you have and simply havent submitted yet or one that you plan to make in the future?
> 
> I ask because I've also run into the need to rename a few packages in a different project of mine, and I've got some ideas and WIP code on how to do it, so I think it makes sense to avoid doubling up work.

Try this twice:

```
CREW_TESTING_REPO=https://github.com/satmandu/chromebrew.git CREW_TESTING_BRANCH=installer_fix CREW_TESTING=1 crew update
```

**Comment by Zopolis4** [comment]
That's the PR for the path fixes, though.

I was asking about the changes for detecting package renames-- I've got a rough idea of how to do it already, so I can follow that through if you want, or you could handle it yourself.

**Comment by satmandu** [comment]
> That's the PR for the path fixes, though.
> 
> I was asking about the changes for detecting package renames-- I've got a rough idea of how to do it already, so I can follow that through if you want, or you could handle it yourself.

Let's discuss! I was thinking of something like thispseudocode (at least in terms of a data structure) inside a refresh_package.rb just like the fixup.rb we just merged, so it could be called from crew after crew update does the git pull...: this doesn't obviously go into what to do once we detect a package name change...

```ruby
def refresh_package_json
  # This could be in an external file.
  pkg_updates = [['meson','mesonbuild',nil], ['qtbase','qt5_base',nil], ['qtchooser',nil,true]]
  array_of_pkg_updates = pkg_updates.collect{|each|Hash[%w{pkg_name new_pkg_name deprecated}.zip(each)]}

  p array_of_pkg_updates
  @device[:installed_packages].each do |package|
    p package
  end
end
```

**Comment by Zopolis4** [comment]
I'll do you one better-- we can get the list of renamed packages from git when updating, no need for a hardcoded list:

```
ami@unbound ~/D/c/lib (vcheck)> git diff HEAD~3 --diff-filter=R --stat
 packages/{acli.rb => acquia_cli.rb}      | 2 +-
 packages/{agrind.rb => angle_grinder.rb} | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)
```

It'll take some sed to get the information, and I still need to figure out the exact command to integrate it into the crew update process, but just posting this as a proof of concept.

I'll post some pseudocode (or maybe even actual code) for a proper integration in a bit.

**Comment by satmandu** [comment]
Git doesn't always detect the renames though... Just look at this PR...

**Comment by Zopolis4** [comment]
Can't you manually tell git that a file was renamed with `git mv`?

**Comment by satmandu** [comment]
> Can't you manually tell git that a file was renamed with `git mv`?

Can you do that for commits far in the past? Wouldn't it be safer to have a master list of renamed packages? (You could always use a git command to help populate such a list...)

**Comment by satmandu** [comment]
@Zopolis4 See what I added to `lib/fixup.rb`...

File: packages/qtcreator.rb [hunk 1]
```diff
@@ -3,14 +3,17 @@
 class Qtcreator < Package
   description 'Qt is a comprehensive cross-platform framework and toolkit that helps you create and build native applications and user interfaces for all the screens of your end user.'
   homepage 'https://info.qt.io/download-qt-for-application-development'
-  version '4.9.1'
-  license '' # Can't find license
-  compatibility 'i686,x86_64'
-  source_url 'https://download.qt.io/official_releases/qt/5.15/5.15.1/md5sums.txt'
-  source_sha256 '50ea8bf9728b8ceb3a92044745edd55fd02783b8733da7fd57e044bc21980bf6'
+  version '4.6.1'
```
Comment (uberhacker) [inline packages/qtcreator.rb:6]: The 4.6.1 installer actually installs 11.0.3.

File: packages/qtcreator.rb [hunk 2]
```diff
@@ -56,22 +59,12 @@ def self.install
...
-    system "mkdir -p #{CREW_DEST_PREFIX}/bin"
-    system "mkdir -p #{CREW_DEST_PREFIX}/share"
+    downloader 'https://download.qt.io/archive/online_installers/4.6/qt-unified-linux-x64-4.6.1-online.run', '11f8e9307a31d283b7ad80050e9d43c1e94dbc2ea283eb6fa9d051549f2bbccb'
+    FileUtils.install 'qt-unified-linux-x64-4.6.1-online.run', "#{CREW_PREFIX}/tmp/qt-unified-linux-x64-4.6.1-online.run", mode: 0o755
+    system "#{CREW_PREFIX}/tmp/qt-unified-linux-x64-4.6.1-online.run --script qt-installer-script.qs"
+    FileUtils.rm_f "#{CREW_PREFIX}/tmp/qt-unified-linux-x64-4.6.1-online.run"
+    FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/bin"
+    FIleUtils.mkdir_p "#{CREW_DEST_PREFIX}/share"
     system "cp -r #{CREW_PREFIX}/share/qt #{CREW_DEST_PREFIX}/share"
     system "ln -s #{CREW_PREFIX}/share/qt/Tools/QtCreator/bin/qtcreator #{CREW_DEST_PREFIX}/bin/qtcreator"
```
Comment (uberhacker) [inline packages/qtcreator.rb:69]: Use FileUtils for the 2 lines above?

Comment (satmandu) [reply packages/qtcreator.rb]: Done.

**Comment by Zopolis4** [comment]
Fair enough. However, can you add a date to each of the fixups in `lib/fixup.rb`? (Including the dates of individual package renames). I think it'd make sense to remove them maybe 2 or 3 years after they were added, and it'd be easier to keep track if the dates are right there in the file.

**Comment by satmandu** [comment]
> Fair enough. However, can you add a date to each of the fixups in `lib/fixup.rb`? (Including the dates of individual package renames). I think it'd make sense to remove them maybe 2 or 3 years after they were added, and it'd be easier to keep track if the dates are right there in the file.

Yes we should definitely add a date and/or a comment/rationale field for that array of changes?

File: packages/qtcreator.rb [hunk 3]
```diff
@@ -56,23 +59,13 @@ def self.install
...
-    system "mkdir -p #{CREW_DEST_PREFIX}/share"
-    system "cp -r #{CREW_PREFIX}/share/qt #{CREW_DEST_PREFIX}/share"
-    system "ln -s #{CREW_PREFIX}/share/qt/Tools/QtCreator/bin/qtcreator #{CREW_DEST_PREFIX}/bin/qtcreator"
+    downloader 'https://download.qt.io/archive/online_installers/4.6/qt-unified-linux-x64-4.6.1-online.run', '11f8e9307a31d283b7ad80050e9d43c1e94dbc2ea283eb6fa9d051549f2bbccb'
+    FileUtils.install 'qt-unified-linux-x64-4.6.1-online.run', "#{CREW_PREFIX}/tmp/qt-unified-linux-x64-4.6.1-online.run", mode: 0o755
+    system "#{CREW_PREFIX}/tmp/qt-unified-linux-x64-4.6.1-online.run --script qt-installer-script.qs"
+    FileUtils.rm_f "#{CREW_PREFIX}/tmp/qt-unified-linux-x64-4.6.1-online.run"
+    FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/bin"
+    FileUtils.mkdir_p "#{CREW_DEST_PREFIX}/share"
+    FileUtils.cp_r "#{CREW_PREFIX}/share/", "#{CREW_DEST_PREFIX}/share"
```
Comment (uberhacker) [inline packages/qtcreator.rb:68]: Won't this result in `#{CREW_DEST_PREFIX}/share/share`?

Comment (uberhacker) [reply packages/qtcreator.rb:68]: Line 67 is not needed with `FileUtils.cp_r "#{CREW_PREFIX}/share/", CREW_DEST_PREFIX.to_s`?

File: lib/fixup.rb [hunk 4]
```diff
@@ -12,3 +12,104 @@
...
+    # maybe put this in a function? e.g.,
+    # rename_pkg pkg[:pkg_name] pkg[:pkg_rename]
+    old_filelist = File.join(CREW_META_PATH, "#{pkg[:pkg_name]}.filelist")
+    new_filelist = File.join(CREW_META_PATH, "#{pkg[:pkg_rename]}.filelist")
+    old_directorylist = File.join(CREW_META_PATH, "#{pkg[:pkg_name]}.directorylist")
+    new_directorylist = File.join(CREW_META_PATH, "#{pkg[:pkg_rename]}.directorylist")
+    # Handle case of new package already installed.
+    if @device[:installed_packages].any? { |elem| elem[:name] == pkg[:pkg_rename] }
+      puts "Renamed #{pkg[:pkg_rename]} is already installed. Deleting old package (#{pkg[:pkg_rename]}) information...".lightblue
+      FileUtils.rm old_filelist if File.file?(old_filelist)
```
Comment (supechicken) [inline lib/fixup.rb:67]: Use `FileUtils.rm_f` (without `File.file?`)?

File: lib/fixup.rb [hunk 5]
```diff
@@ -12,3 +12,104 @@
...
+end
+require 'active_support/core_ext/object/blank'
+
+pkg_update_arr.each do |pkg|
+  next unless @device[:installed_packages].any? { |elem| elem[:name] == pkg[:pkg_name] }
+
+  puts "#{pkg[:pkg_name]} found in package fixup list".orange
+
+  # Package rename.
+  unless pkg[:pkg_rename].blank?
```
Comment (supechicken) [inline lib/fixup.rb:56]: Use `.to_s.empty?` instead of external gems?

File: lib/fixup.rb [hunk 6]
```diff
@@ -12,3 +12,104 @@
...
+    # If new filelist or directorylist do not exist and new package is not
+    # marked as installed in device.json then rename and edit device.json .
+    puts "Renaming #{pkg[:pkg_name]} to #{pkg[:pkg_rename]}".lightblue
+    begin
+      FileUtils.cp "#{CREW_CONFIG_PATH}/device.json", "#{CREW_CONFIG_PATH}/device.json.bak"
+      FileUtils.mv old_filelist, new_filelist
+      FileUtils.mv old_directorylist, new_directorylist
+      # This following line isn't working to rename inside the data
+      # structure pulled from the json file.
+      @device[:installed_packages].map { |x| x[:name] == pkg[:pkg_name] ? pkg[:pkg_rename] : x[:name] }
```
Comment (supechicken) [inline lib/fixup.rb:91]: I think you mean this:
```ruby
@device[:installed_packages].map do |x|
  x[:name] = pkg[:pkg_rename] if x[:name] == pkg[:pkg_name]
  next x
end
```

Comment (satmandu) [reply lib/fixup.rb:91]: OMFG I added a `=`...

Comment (satmandu) [reply lib/fixup.rb]: Thanks for this... I was tearing my hair out over this, metaphorically speaking...

**Comment by satmandu** [comment]
This is pretty great:

image

File: lib/fixup.rb (hunk 5 above)
Comment (satmandu) [reply lib/fixup.rb]: Done!

File: packages/qtcreator.rb (hunk 3 above)
Comment (satmandu) [reply packages/qtcreator.rb]: done!

**Comment by satmandu** [comment]
Just waiting for the arm build of qt5_webengine to finish...

File: lib/fixup.rb [hunk 7]
```diff
@@ -12,3 +12,115 @@
...
+  { pkg_name: 'qtwebsockets', pkg_rename: 'qt5_websockets', pkg_deprecated: nil, comments: nil },
+  { pkg_name: 'qtx11extras', pkg_rename: 'qt5_x11extras', pkg_deprecated: nil, comments: nil },
+  { pkg_name: 'qtchooser', pkg_rename: nil, pkg_deprecated: true, comments: "Doesn't work for newer Qt versions." }
+]
+
+pkg_update_arr.each do |pkg|
+  next unless @device[:installed_packages].any? { |elem| elem[:name] == pkg[:pkg_name] }
+
+  puts
+  puts "#{pkg[:pkg_name].capitalize} found in package fixup list".lightcyan
```
Comment (uberhacker) [inline lib/fixup.rb:45]: Line 44 could be removed with `puts "\n#pkg...`.

Comment (satmandu) [reply lib/fixup.rb]: Changed.

File: packages/qt5_charts.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  depends_on 'qt5_declarative' # R
+
+  def self.build
+    system 'qmake'
+    system 'make'
+  end
+
+  def self.install
+    system "make INSTALL_ROOT=#{CREW_DEST_DIR} install"
+  end
```
Comment (uberhacker) [inline packages/qt5_charts.rb:36]: Should we add the build and install sections to `buildsystems/qmake`?

File: lib/fixup.rb [hunk 9]
```diff
@@ -12,3 +12,114 @@
...
+      File.write "#{CREW_CONFIG_PATH}/device.json", JSON.pretty_generate(JSON.parse(@device.to_json))
+      puts "#{pkg[:pkg_name].capitalize} renamed to #{pkg[:pkg_rename].capitalize}".lightgreen
+    rescue StandardError => e
+      puts 'Restoring old filelist, directorylist, and device.json...'.lightred
+      FileUtils.mv new_filelist, old_filelist
+      FileUtils.mv new_directorylist, old_directorylist
+      FileUtils.cp "#{CREW_CONFIG_PATH}/device.json.bak", "#{CREW_CONFIG_PATH}/device.json"
+      # Reload json file.
+      @device = JSON.load_file("#{CREW_CONFIG_PATH}/device.json", symbolize_names: true)
+      @device.transform_values! {|val| val.is_a?(String) ? val.to_sym : val }
```
[the remark this answers is not in the document]
Comment (satmandu) [reply lib/fixup.rb:97]: Good call.

File: packages/qt5_charts.rb (hunk 8 above)
Comment (satmandu) [reply packages/qt5_charts.rb:36]: I was thinking about it...

**Review by uberhacker — APPROVED** [review]

**Pull request merged**